Add charts to store and node pages.

BramGruneir opened the pull request:
Adds:
- Stats and Charts on the store/:store_id page
- Stats and Charts on the node/:node_id pages
- Removes dead links from the header.

This is still rough, but I wanted to get something in for the demo tomorrow.  Goal for this PR was purely functionality.

Note that I couldn't get the byte charts working as there is a bug in the line chart code.  I'll try to fix that tomorrow.

Future work to be done:
- There's going to be lots of work refactoring the code and adding all the missing comments.
- I'd like to batch the queries together instead of having each one be a different network call,
- Lots of design work to make the pages look nicer.
- More work on the charts.
- I'm going to add the statistics table to the node and store pages.  This should be easy to do, I'll try to get that in tomorrow as well. --- This is now included in this PR.

Store Page 1/2:
![screen shot 2015-05-28 at 03 32 42](https://cloud.githubusercontent.com/assets/1614265/7854763/673718aa-04ea-11e5-8d82-419f93c34574.png)

Store Page 2/2:
![screen shot 2015-05-28 at 03 32 58](https://cloud.githubusercontent.com/assets/1614265/7854762/67357ca2-04ea-11e5-93b6-2cf6f24a9053.png)

Node Page 1/2:
![screen shot 2015-05-28 at 03 33 27](https://cloud.githubusercontent.com/assets/1614265/7854765/673ac0c2-04ea-11e5-8ca6-5a8647a076a1.png)

Node Page 2/2:
![screen shot 2015-05-28 at 03 33 36](https://cloud.githubusercontent.com/assets/1614265/7854764/6738b336-04ea-11e5-883b-ca8ced2f9d83.png)

BramGruneir committed:
Add charts to store and node pages.

petermattis commented:
The "Successful Calls" and "Error Calls" graphs are not too useful. I'd dump them and only keep the associated "Rate" graphs.

BramGruneir commented:
I can drop those two.  How's it looking otherwise?

petermattis commented:
It look functional for the demo. We're going to get design help to make it pretty in the coming weeks.

I noticed the tooltips on your graphs are nowhere near where the graph is. Bug?

BramGruneir commented:
Great.
And yeah, that's another bug, they should appear near the mouse.  I'm going to see if I can get that fixed as well.

BramGruneir committed:
Add nodes and stores details and stats.

BramGruneir commented:
Removed the success and error chart from the node detail page.
Added all nodes and stores statistics to the listing pages.

Note that all values are polled every 10s from the webpage.

![screen shot 2015-05-28 at 11 54 16](https://cloud.githubusercontent.com/assets/1614265/7864489/5cd6ba46-0530-11e5-9adc-61912f48c342.png)
![screen shot 2015-05-28 at 11 54 25](https://cloud.githubusercontent.com/assets/1614265/7864490/5cd8e168-0530-11e5-82ee-8a8616f3bd4e.png)

tamird commented on resource/ui/ts/models/node_status.ts:
```diff
@@ -141,10 +139,70 @@ module Models {
+                    }
+                };
+
+                var nodeId: string;
```
do you really need this declaration outside the `for (...) ` below/

BramGruneir replied:
I tried putting it in the for with a 
for (var nodeID in ...
But it kept failing.  I'll try it again.

BramGruneir replied:
Found the issue.  Fixed.

tamird commented on resource/ui/ts/models/node_status.ts:
```diff
@@ -141,10 +139,70 @@ module Models {
                         m("tr", [m("td", "Leader Ranges:"), m("td", node.leader_range_count)]),
                         m("tr", [m("td", "Available Ranges:"), m("td", node.available_range_count)]),
-                        m("tr", [m("td", "Availablility:"), m("td", this._availability(nodeId))]),
+                        m("tr", [m("td", "Availablility:"), m("td", Nodes._availability(node))]),
```
wouldn't `this` still work? more generally: what is the value of `this` in a typescript static method?

BramGruneir replied:
Static method exists on the class itself, so you have to reference it using the classname.
`this` in a static method should refer to the class itself, not the instance (since you wouldn't have one).  But with JS, the context can get mixed up quickly.  See the func.call(instance) that I have to do in the refresher.

tamird replied:
Whoa, that is bonkers.

> this in a static method should refer to the class itself

What are describing here? Not ts/js behaviour, I take it?

BramGruneir replied:
So this was interesting so I wanted to test it out.
When in a static method, the `this` refers to the class (or to be more exact, the function that is the definition of the class)
`
this
Stores()
`
referes to:
`
 var Stores = (function () {
            function Stores() {
                this._data = m.prop({});
                this.desc = m.prop({});
                this.statuses = m.prop({});
            }
            Stores.prototype.Query = function () {
  ...`

BramGruneir replied:
So in the end, when in a static function, don't use the `this`.  Ever.

tamird replied:
wait, what's at the end of those ellipsis? if `this` is just a value, why not use it?

BramGruneir replied:
That's just from the generated code.  I just cut out a section of it and added the ellipsis myself.
this = Stores() -- when you're inside a static function.
When you're in an inside a non-static function, `this` points to the instance of the class.  You cannot access any static functions from the `this` pointer if you're inside an instance. since the functions are not on the instance.

tamird replied:
right, but in this case you're inside a static method, so `this` should be the class instance?

tamird replied:
oh, nm, this is not a static method. Ignore me and carry on! though, something akin to ruby's  `self.class` would be neat.

BramGruneir replied:
Agreed, self.class would be great.
And even if we were in a static function, the this doesn't seem to give you access to the other static elements, it's just a function representing the class.  Very strange.  I tried chaining things off of it, and it wouldn't let me in the debugger.

tamird commented on resource/ui/ts/models/node_status.ts:
```diff
@@ -141,10 +139,70 @@ module Models {
+                    if (nodeStatus.updated_at > status.updated_at) {
+                        status.updated_at = nodeStatus.updated_at;
+                    }
+                    status.stats.live_bytes += nodeStatus.stats.live_bytes;
```
this field-by-field addition seems fraught with danger. is it possible to implement the addition operator in TS? might be good to do in a future PR.

BramGruneir replied:
Yeah, this is rather ugly.  I'll add a todo about a refactoring here.

BramGruneir replied:
done.

tamird commented on resource/ui/ts/models/stats.ts:
```diff
@@ -22,5 +22,61 @@ module Models {
+        // This function was adapted from
+        // https://stackoverflow.com/questions/10420352/converting-file-size-in-bytes-to-human-readable
+        export function FormatBytes(bytes: number): string {
+            var thresh = 1024;
```
does ts have a notion of constants?

tamird replied:
also perhaps a better name is `kilo`?

tamird replied:
`kibi`, ratehr

BramGruneir replied:
TS doesn't seem to have constants. I'll pull it out of the function at least.  And yes, I'll change the name to kibi. :)

tamird commented on resource/ui/ts/models/stats.ts:
```diff
@@ -22,5 +22,61 @@ module Models {
+            if (Math.abs(bytes) < thresh) {
+                return bytes + ' B';
+            }
+            var units = ['KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB', 'ZiB', 'YiB'];
```
nit: you can simplify this function by prepending `'B'` to this array (and making it a constant) and indexing into it by `log2(bytes)/10`.

[log2](https://github.com/Microsoft/TypeScript/blob/master/src/lib/es6.d.ts#L541

BramGruneir replied:
Converted to constants.  I'll update the function using log2 in another PR.

tamird commented on resource/ui/ts/models/store_status.ts:
```diff
@@ -136,10 +134,69 @@ module Models {
+                    status.stats.live_bytes += storeStatus.stats.live_bytes;
+                    status.stats.key_bytes += storeStatus.stats.key_bytes;
+                    status.stats.val_bytes += storeStatus.stats.val_bytes;
+                    status.stats.intent_bytes += storeStatus.stats.intent_bytes;
```
This looks like the same code from node_status. Add a TODO to extract?

BramGruneir replied:
done.

tamird commented on resource/ui/ts/pages/nodes.ts:
```diff
@@ -13,14 +15,66 @@ module AdminViews {
+        }
+      }
+
+      private static _queryManagerBuilder(nodeId: string, agg: Models.Metrics.QueryAggregator, source:string): Models.Metrics.QueryManager {
```
i dont meaningfully understand this code, so not reviewing it in detail. nit: space after `source:`

BramGruneir replied:
This is the fancy JS I had to come up with to get multiple charts working in a sane manner.  
The builder creates instances of the queryManager (each one essentially represents a single chart on a page at this point.)
The _refresh function is where all the requery logic sits, to update the data every interval.

tamird commented on resource/ui/ts/typings/mithriljs/mithril.d.ts:
```diff
@@ -122,8 +122,8 @@ declare module _mithril {
 		toJSON(): T;
 	}
 
-	interface MithrilRoutes<T extends MithrilController> {
```
why these changes?

BramGruneir replied:
The typescript typings were incorrect.  Matt was the one who actually did the work on this.  If we don't do this, the transpiler fails when it tries to sort out the routing table in app.ts.

tamird replied:
K.

BramGruneir committed:
Updates based on code review.

BramGruneir commented:
updated, @tamird PTAL

tamird commented on resource/ui/ts/models/stats.ts:
```diff
@@ -22,5 +22,61 @@ module Models {
+
+        // This function was adapted from
+        // https://stackoverflow.com/questions/10420352/converting-file-size-in-bytes-to-human-readable
+        var kibi = 1024
```
did you see my other comment about using log?

BramGruneir replied:
yep, responded to it.

tamird commented:
I don't want to block you, so LGTM. It'd be nice to address the line of discussion on the use of `this` and to simplify `FormatBytes` using `log2`.

BramGruneir commented:
I think I've addressed the `this` usage sufficiently.  
I'll convert FormatBytes to use log2 in another PR.

tamird commented on resource/ui/ts/models/node_status.ts:
```diff
@@ -78,38 +78,34 @@ module Models {
                 this.desc(<NodeDescriptionMap> {});
-                var nodeId: string;
-                for (nodeId in this._data()) {
+                for (var nodeId in this._data()) {
```
neat
